src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php

```diff
@@ -0,0 +1,85 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel\EventListener;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpKernel\Event\FinishRequestEvent;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+use Symfony\Contracts\Service\LocaleAwareInterface;
+use TypeError;
+
+/**
+ * Inject request locale to each services that implements LocaleAwareInterface.
+ *
+ * @author Pierre Bobiet <pierrebobiet@gmail.com>
+ */
+class LocaleAwareListener implements EventSubscriberInterface
+{
+    /**
+     * @var \Symfony\Contracts\Service\LocaleAwareInterface[]
+     */
+    private $localeAwareServices;
+    private $requestStack;
+
+    public function __construct(iterable $localeAwareServices, RequestStack $requestStack)
+    {
+        $this->localeAwareServices = $localeAwareServices;
+        $this->requestStack = $requestStack;
+    }
+
+    public function onKernelRequest(GetResponseEvent $event): void
+    {
+        $this->setLocale($event->getRequest());
+    }
+
+    public function onKernelFinishRequest(FinishRequestEvent $event): void
+    {
+        if (null === $parentRequest = $this->requestStack->getParentRequest()) {
+            return;
+        }
+
+        $this->setLocale($parentRequest);
+    }
+
+    public static function getSubscribedEvents()
+    {
+        return array(
+            // must be registered after the Locale listener
+            KernelEvents::REQUEST => array(array('onKernelRequest', 10)),
+            KernelEvents::FINISH_REQUEST => array(array('onKernelFinishRequest', 0)),
+        );
+    }
+
+    private function setLocale(Request $request): void
+    {
+        foreach ($this->localeAwareServices as $i => $service) {
+            if (!$service instanceof LocaleAwareInterface) {
+                throw new TypeError(sprintf(
+                    'Value #%d of argument 1 passed to %s must be an instance of %s, %s given',
```

Comment:
locale aware services are not an argument of `__METHOD__` here.

something like `Service must be an instance of "%s", got "%s".` would be more accurate :) but perhaps we should find a way to move the check to compile time, e.g.

```xml
<argument type="tagged" tag="locale_aware" instanceof="SF\...\LocaleAwareInterface"/>
```